### ISSUE TITLE
[Regression] Segfault with backward pass in DiffTensorView

### ISSUE THREAD

**Issue by mkeshavaNV** [shader-slang/slang#10417]
# Issue Description

`bwd_diff` through a `[Differentiable]` extension method on `DiffTensorView` that uses a `[ForceUnroll]` loop with index mutation (`x[R] += L`) produces a segfault at runtime when the method calls another `[Differentiable]` extension method in the loop body.

Specifically, `storeMatrixOnce` (loops over rows calling `storeVecOnce` with `x[R] += L`) and `loadMatrixOnce` (loops over rows calling `loadVecOnce` with `x[R] += L`) crash in the backward pass. The manually unrolled equivalent (calling `storeVecOnce`/`loadVecOnce` directly with fixed offsets) works correctly.

This is a regression: it works in slangc 2026.2.1 and segfaults in slangc 2026.3.1.

# Reproducer Code

**repro.slang:**

```slang
// Extension methods on DiffTensorView

__generic<T : __BuiltinFloatingPointType>
extension DiffTensorView<T> {

    // Vector load/store — backward works fine
    [Differentiable]
    __generic<let M : int, let R : int, let N : int>
    vector<T, M> loadVecOnce(vector<uint, N> x) {
        vector<T, M> result;
        [ForceUnroll]
        for (int j = 0; j < M; j++) {
            result[j] = this.loadOnce(x);
            x[R] += 1;
        }
        return result;
    }

    [Differentiable]
    __generic<let M : int, let N : int>
    vector<T, M> loadVecOnce(vector<uint, N> x) {
        return this.loadVecOnce<M, N - 1, N>(x);
    }

    [Differentiable]
    __generic<let M : int, let R : int, let N : int>
    void storeVecOnce(vector<uint, N> x, vector<T, M> y) {
        [ForceUnroll]
        for (int j = 0; j < M; j++) {
            this.storeOnce(x, y[j]);
            x[R] += 1;
        }
    }

    [Differentiable]
    __generic<let M : int, let N : int>
    void storeVecOnce(vector<uint, N> x, vector<T, M> y) {
        this.storeVecOnce<M, N - 1, N>(x, y);
    }

    // Matrix store — backward SEGFAULTS in 2026.3.1
    [Differentiable]
    __generic<let K : int, let L : int, let R : int, let N : int>
    void storeMatrixOnce(vector<uint, N> x, matrix<T, K, L> value) {
        [ForceUnroll]
        for (int i = 0; i < K; i++) {
            this.storeVecOnce<L, R, N>(x, value[i]);
            x[R] += L;
        }
    }

    [Differentiable]
    __generic<let K : int, let L : int, let N : int>
    void storeMatrixOnce(vector<uint, N> x, matrix<T, K, L> value) {
        this.storeMatrixOnce<K, L, N - 1, N>(x, value);
    }

    // Matrix load — backward SEGFAULTS in 2026.3.1
    [Differentiable]
    __generic<let K : int, let L : int, let R : int, let N : int>
    matrix<T, K, L> loadMatrixOnce(vector<uint, N> x) {
        matrix<T, K, L> result;
        [ForceUnroll]
        for (int i = 0; i < K; i++) {
            result[i] = this.loadVecOnce<L, R, N>(x);
            x[R] += L;
        }
        return result;
    }

    [Differentiable]
    __generic<let K : int, let L : int, let N : int>
    matrix<T, K, L> loadMatrixOnce(vector<uint, N> x) {
        return this.loadMatrixOnce<K, L, N - 1, N>(x);
    }
};

// SEGFAULTS in backward with slangc 2026.3.1:
[CUDAKernel]
[Differentiable]
[AutoPyBindCUDA]
void store_matrix_kernel(
    DiffTensorView translation, // (N, 3)
    DiffTensorView rotation,    // (N, 4)
    DiffTensorView result       // (N, 16)
) {
    uint idx = cudaThreadIdx().x + cudaBlockIdx().x * cudaBlockDim().x;
    if (idx >= translation.size(0)) return;

    float3 trans = translation.loadVecOnce<3>(uint2(idx, 0));
    float4 rot = rotation.loadVecOnce<4>(uint2(idx, 0));

    float4x4 mat;
    mat[0] = float4(rot.x, rot.y, rot.z, 0.0);
    mat[1] = float4(rot.y, rot.w, rot.x, 0.0);
    mat[2] = float4(rot.z, rot.x, rot.w, 0.0);
    mat[3] = float4(trans.x, trans.y, trans.z, 1.0);

    result.storeMatrixOnce(uint2(idx, 0), mat);  // <-- backward segfaults
}

// WORKS in backward — same logic, manually unrolled:
[CUDAKernel]
[Differentiable]
[AutoPyBindCUDA]
void store_matrix_manual_kernel(
    DiffTensorView translation, // (N, 3)
    DiffTensorView rotation,    // (N, 4)
    DiffTensorView result       // (N, 16)
) {
    uint idx = cudaThreadIdx().x + cudaBlockIdx().x * cudaBlockDim().x;
    if (idx >= translation.size(0)) return;

    float3 trans = translation.loadVecOnce<3>(uint2(idx, 0));
    float4 rot = rotation.loadVecOnce<4>(uint2(idx, 0));

    float4x4 mat;
    mat[0] = float4(rot.x, rot.y, rot.z, 0.0);
    mat[1] = float4(rot.y, rot.w, rot.x, 0.0);
    mat[2] = float4(rot.z, rot.x, rot.w, 0.0);
    mat[3] = float4(trans.x, trans.y, trans.z, 1.0);

    // Manually unrolled — backward works fine:
    result.storeVecOnce(uint2(idx, 0), mat[0]);
    result.storeVecOnce(uint2(idx, 4), mat[1]);
    result.storeVecOnce(uint2(idx, 8), mat[2]);
    result.storeVecOnce(uint2(idx, 12), mat[3]);
}
```

**repro.py** (requires `slangpy` and `torch`):

```python
import os
import torch
import slangpy as spy
import slangpy.torchintegration.torchtensormarshall  # registers torch.Tensor handler
from slangpy.torchintegration import diff_pair

torch.cuda.init()
handles = spy.get_cuda_current_context_native_handles()
device = spy.Device(
    type=spy.DeviceType.cuda,
    compiler_options={
        "include_paths": [],
        "defines": {"SLANGPY_DISPATCH": "1"},
        "optimization": spy.SlangOptimizationLevel.maximal,
        "floating_point_mode": spy.SlangFloatingPointMode.fast,
    },
    enable_print=True,
    enable_hot_reload=False,
    existing_device_handles=handles,
)

slang_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), "repro.slang")
module = spy.Module.load_from_file(device, slang_file)

N = 4
t = torch.randn(N, 3, device="cuda")
r = torch.randn(N, 4, device="cuda")

# Test 1: manually unrolled — PASSES
print("Test 1: store_matrix_manual_kernel backward...")
res = torch.zeros(N, 16, device="cuda")
module.store_matrix_manual_kernel(_thread_count=N, translation=t, rotation=r, result=res)
gt, gr, gres = torch.zeros_like(t), torch.zeros_like(r), torch.ones_like(res)
module.store_matrix_manual_kernel.bwds(
    _thread_count=N,
    translation=diff_pair(t, gt), rotation=diff_pair(r, gr), result=diff_pair(res, gres))
print(f"  PASSED (grad={gt[0].tolist()})")

# Test 2: storeMatrixOnce — SEGFAULTS with slangc 2026.3.1
print("Test 2: store_matrix_kernel backward (storeMatrixOnce)...")
res2 = torch.zeros(N, 16, device="cuda")
module.store_matrix_kernel(_thread_count=N, translation=t.clone(), rotation=r.clone(), result=res2)
gt2, gr2, gres2 = torch.zeros_like(t), torch.zeros_like(r), torch.ones_like(res2)
print("  Running backward...", flush=True)
module.store_matrix_kernel.bwds(
    _thread_count=N,
    translation=diff_pair(t.clone(), gt2), rotation=diff_pair(r.clone(), gr2),
    result=diff_pair(res2, gres2))
print(f"  PASSED (grad={gt2[0].tolist()})")
```

# Expected Behavior

Both `store_matrix_kernel` and `store_matrix_manual_kernel` should produce correct gradients in the backward pass. `storeMatrixOnce` is semantically equivalent to manually calling `storeVecOnce` four times with fixed offsets — both should work identically.

# Actual Behavior

With slangc 2026.3.1, calling `.bwds()` on `store_matrix_kernel` (which uses `storeMatrixOnce`) produces a segfault:

```
Fatal Python error: Segmentation fault

Current thread 0x00007847cbfff6c0 (most recent call first):
  File "slangpy/core/function.py", line 287 in call    # _native_call
```

The manually unrolled `store_matrix_manual_kernel` works correctly with the same compiler version.

With slangc 2026.2.1, both kernels work correctly — no segfault.

# Environment

- **Slang Version**: Regression between 2026.2.1 (works) and 2026.3.1 (segfaults)
- **OS**: Linux (Ubuntu), kernel 6.14.0-35-generic, x86_64
- **slangpy version**: 0.40.1
- **PyTorch**: 2.7.0
- **Python**: 3.11

**Comment by jhelferty-nv** [comment]
@mkeshavaNV Can you try root-causing with an agent? Once we have a root cause report, if the agent can't solve it itself, Sai can take a look.

**Comment by jhelferty-nv** [comment]
## Root Cause Analysis

The bug has two contributing factors that combine to produce invalid CUDA code in the backward pass.

### 1. `eliminatePhis` pass preserves `@ConstExpr` rate on mutable variables

When the `eliminatePhis` pass converts SSA `updateElement` instructions into mutable variable operations (var + getElementPtr + store), it preserves the `@ConstExpr` rate qualification from the original instruction. For example, `uint2 idx = uint2(0, 0)` is constant-initialized, so the `updateElement` that represents `idx.y += 4` carries `@ConstExpr`. The resulting temp var gets type `Ptr(RateQualified(ConstExpr, Vec(UInt, 2)))` instead of `Ptr(Vec(UInt, 2))`.

### 2. CPP/CUDA emitter doesn't unwrap `RateQualified` when checking for vector types

The CPP emitter has a special handler for `kIROp_GetElementPtr` on vector types (`slang-emit-cpp.cpp`, `tryEmitInstExprImpl`). For vectors with a static index, it emits valid CUDA like `&(base->y)`. For dynamic indices it emits `_slang_vector_get_element_ptr(...)`. The check is:

```cpp
IRType* baseType = as<IRPtrTypeBase>(baseInst->getDataType())->getValueType();
if (auto vectorBaseType = as<IRVectorType>(baseType))
```

When `baseType` is `RateQualified(ConstExpr, Vec(UInt, 2))`, the `as<IRVectorType>()` cast returns null. The handler returns `false`, falling through to the generic `CLikeSourceEmitter` which emits `base[int(1)]` — invalid CUDA since `uint2` doesn't support `operator[]`.

The same issue exists in `emitDereferenceOperand` in `slang-emit-c-like.cpp`, which also checks `as<IRVectorType>(ptrType->getValueType())` without unwrapping.

### Why the inner loop works but the outer loop doesn't

In the backward pass, `innerLoop` receives `idx` as a **value parameter** (`Vec(UInt, 2)`), so the temp var created by `eliminatePhis` gets type `Ptr(Vec(UInt, 2))` — no `RateQualified` wrapper. The vector type check succeeds, and the emitter produces valid `*&((&_S3)->y) = _S2`.

In `outerLoop`, `idx` is a **local variable** initialized with a constant (`uint2(0, 0)`), so it carries `@ConstExpr`. The temp var gets `Ptr(RateQualified(ConstExpr, Vec(UInt, 2)))`, the vector check fails, and the emitter produces invalid `_S12[int(1)] = 4U`.

### Minimal reproducer (no slangpy/DiffTensorView needed)

```slang
//TEST(compute):COMPARE_COMPUTE_EX(filecheck-buffer=CHECK):-cuda -compute -shaderobj -output-using-type

//TEST_INPUT:ubuffer(data=[0 0 0 0], stride=4):out,name=outputBuffer
RWStructuredBuffer<float> outputBuffer;

[Differentiable]
float innerLoop(float4 values, uint2 idx)
{
    float result = 0.0;
    [ForceUnroll]
    for (int j = 0; j < 4; j++)
    {
        result += values[j] * float(idx.y + 1);
        idx.y += 1;
    }
    return result;
}

[Differentiable]
float outerLoop(float4x4 mat)
{
    float result = 0.0;
    uint2 idx = uint2(0, 0);
    [ForceUnroll]
    for (int i = 0; i < 4; i++)
    {
        result += innerLoop(mat[i], idx);
        idx.y += 4;
    }
    return result;
}

[numthreads(1, 1, 1)]
void computeMain(uint3 dispatchThreadID: SV_DispatchThreadID)
{
    float4x4 mat = float4x4(
        1.0, 2.0, 3.0, 4.0,
        5.0, 6.0, 7.0, 8.0,
        9.0, 10.0, 11.0, 12.0,
        13.0, 14.0, 15.0, 16.0
    );
    DifferentialPair<float4x4> dp = diffPair(mat, float4x4(0.0));
    __bwd_diff(outerLoop)(dp, 1.0);
    outputBuffer[0] = dp.d[0][0]; // CHECK: 1.0
    outputBuffer[1] = dp.d[1][0]; // CHECK: 5.0
    outputBuffer[2] = dp.d[2][0]; // CHECK: 9.0
    outputBuffer[3] = dp.d[3][0]; // CHECK: 13.0
}
```

CUDA compilation fails with:
```
nvrtc 12.6: error: no operator "[]" matches these operands
            operand types are: uint2 [ int ]
            _S12[int(1)] = 4U;
```

### Suggested Fix

Call `unwrapAttributedType()` before the `as<IRVectorType>()` checks in two places:

1. **`slang-emit-cpp.cpp`** (`tryEmitInstExprImpl`, `kIROp_GetElementPtr` case):
   ```cpp
   IRType* baseType = (IRType*)unwrapAttributedType(
       as<IRPtrTypeBase>(baseInst->getDataType())->getValueType());
   ```

2. **`slang-emit-c-like.cpp`** (`emitDereferenceOperand`, `kIROp_GetElementPtr` case):
   ```cpp
   vectorType = as<IRVectorType>(unwrapAttributedType(ptrType->getValueType()));
   ```

This is consistent with how the address elimination pass already unwraps attributed types before checking for `IRNonCopyableTypeDecoration` in `slang-ir-addr-inst-elimination.cpp`.
